### ISSUE TITLE
Sunflower copies hackeds liquidity pool smart contract

### ISSUE THREAD

**Issue by look4mexpz** [sunflower-land/sunflower-land#252]
Hello my dears.

I can't stand to see these games being hacked anymore. Sunflower was open source, and with that many people are using its source codes and creating version variations of the games, but it has a major flaw that all of them, or at least (almost all), are being hacked, being stolen from liquidity pool somehow.

Someone with more experience in blockchain contracts would be able to tell where this flaw is, which allows this hacker to run this exploit.

I will leave below the two wallets that the hacker uses and some of his actions that he performed to steal the liquidity pool.

[![roub2.png](https://i.postimg.cc/9QNjhm6b/roub2.png)](https://postimg.cc/qNnFcTM6)
HFF - 0x36c0845773a6066a25f228e62248915ef481c38c

[![ROUB1.png](https://i.postimg.cc/jjwYdVKb/ROUB1.png)](https://postimg.cc/7CkWmWfs)
CFG - 0xa1F8541DF23Cf0F3551BDc665E9Abd31f125Bc37

Above are 2 heists on 2 different tokens.

Hacker Address 1: 0x5750477413ba057e068d95f7043a313dc8a26c10
Hacker Address 2: 0x6c9244d2f9febf230f7be9de2d185b330e431261

TxId Hacker 1.1: https://polygonscan.com/tx/0xb4127ff702af9a96ff97c0c01d76f9966984b80cb63fec352e76d611e3120b7e
TxId Hacker 1.2: https://polygonscan.com/tx/0xe0c0621b54cd7cb63bba2fb19913b00d644e3f003608111d7fe00aee6f8cdef6
TxId Hacker 2.1: https://bscscan.com/tx/0x6c4e018f6f427108695f029237828704aec17799020837236fbf46c475efeef4

TxId Hacker 2.2: https://bscscan.com/tx/0xe89a289c56f1477311d1e17189df764cc5791cae6032e0547458a3698f37e909

TxId Hacker 3.1: https://polygonscan.com/tx/0x50cd0dd569d45ac9b286d8deb5b9eb72bd0b8121c7ef524a5d6a2e270af0402e

TxId Hacker 3.2: https://polygonscan.com/tx/0x523ea53d8deaa0a726825739e0bf4e81a1c9c83b45cfd8000fe4d22a3663c8cd

Attention: https://polygonscan.com/token/0x1816a91ee2e30aa39dee4fd24590faf0087b5a6c?a=0x1b92f51d46608a8d052acb4e556c61e6703e12e9

I will list below all the versions that I know of and if they were stolen or not.

SUNFLOWER - POLYGON - (ORIGINAL) (NOT HACKED)
TOKEN: 0xdf9b4b57865b403e08c85568442f95c26b7896b0

FRUIT FARM - BSC - (NOT HACKED)
TOKEN: 0xa2e05fee995d84e388111065f9da0e1fd0358a0b

MY LITTLE FARM - POLYGON - (HACKED)
TOKEN: 0xed2f85f446281a31bcae074938e867422f5074c8

FISHLAND - POLYGON - (HACKED)
TOKEN: 0x1816a91ee2e30aa39dee4fd24590faf0087b5a6c

CAT FARM - BSC - (HACKED)
TOKEN: 0xa1F8541DF23Cf0F3551BDc665E9Abd31f125Bc37

HERO FARM - POLYGON - (HACKED)
TOKEN: 0x36c0845773a6066a25f228e62248915ef481c38c

MOONFLOWER - AURORA - (HACKED)
TOKEN: 0x1103f33c4aa272dd3fdd73724f492ee35f097644

CRYPTO FARM TOWN - AURORA - (HACKED)
TOKEN: 0x330db1dd6d150727eeb53feaf5d485bce2aa326f

The only copy that was not hacked was Fruit Farm.

What do you have to do to correct this huge flaw?

Copies have been appearing all over the world, and we need someone who can fix it, because we know it won't stop being released, and that only has to denigrate the image of "crypto", "blockchain", "NFT". With that, fewer people can come to this market, and we can't let this hacker get away with taking liquidity out of these contracts.

Anyone who knows how to fix this problem, let's try to publicize this fix so that people can stop losing money, because many enter these projects, and most of them lose, and yet they think it was rug pull, further staining our nft market .

We know you didn't like your codes being used like that, but think about the people who end up getting into these projects and getting ripped off.

Could you release a fix for this bug, for those who have the old source code.

This would not harm you, as you are working on a V2, much more robust, it would only make this hacker not earning a lot of money, because every day 1 sunflower copies is born.

**Comment by paluras** [comment]
I don't believe this post serves any good to the development of the game. The team is not affiliated whit any of these projects

**Comment by Protocteur** [comment]
i've made some research, and in my opinion that need to be checked by a solidity expert of the team because that could happen to SFL too if i understand well how the hack has been done, because it seems that there is a way to burn token without needing to get any access on the token contract.

How it's seem to happen :

Solidity as a vulnerability associated with the "selfdestruct" method. This method is used to end a contract, making it not usable anymore.

Doing a selfdestruct has 2 consequences :
All ether on the contract are transfered to the adresse used for the selfdestruct
AND all others assets hold by the account are burn.

So in fact you doesn't need to get access to the main contract, just create your own, transfer almost all token you bought to the contract exploit and a small part to another adress. Then you selfdestruct the contract you own with all token in it, and finaly sale the token you keep on the other adress at an overprice price.

But this exploit only work if you can burn enought token to increase the price of the token so high that the new price cover your initial investment, and if you're saling your token just after the burn, because other people will sale what they have too if you didn't stole the whole liquidity.

Not sure if i am right, i'm a fullstack dev, but never used solidity.

To do your own research : 

https://solidity-by-example.org/hacks/self-destruct/
https://betterprogramming.pub/solidity-what-happens-with-selfdestruct-f337fcaa58a7
https://docs.soliditylang.org/en/v0.8.11/introduction-to-smart-contracts.html?highlight=self%20destruct#deactivate-and-self-destruct

--

The documents doesn't explicitly say what happens to tokens/assets, but it's seems the contrat burn them instead of holding them forever. And that where the exploit seems to came from.

**Comment by adamhannigan** [comment]
As stated, the team does not support, endorse or have any relation to these contracts. It is a bit difficult to say what the issue is without diving into the code.

I will leave this issue open for awareness but label it as a discussion.

**Comment by adamhannigan** [comment]
Closing for inactivty
